theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,741 @@
+        * ``int``: pad input with a symmetric border of zeros of the given
+          width, then perform a valid convolution.
+        * ``(int1, int2)``: pad input with a symmetric border of ``int1`` rows
+          and ``int2`` columns, then perform a valid convolution.
+
+    :type subsample: tuple of len 2
+    :param subsample: factor by which to subsample the output.
+        Also called strides elsewhere.
+
+    :type filters_flip: bool
+    :param filters_flip: If ``True``, will flip the filter rows and columns
+        before sliding them over the input. This operation is normally referred
+        to as a convolution, and this is the default. If ``False``, the filters
+        are not flipped and the operation is referred to as a cross-correlation.
+
+    :rtype: symbolic 4D tensor
+    :return: set of feature maps generated by convolutional layer. Tensor is
+        of shape (batch size, output channels, output rows, output columns)
+    """
+
+    ### FIXME input shape/kernel shape
+    conv_op = AbstractConv2d(imshp=inputs_shape,
+                             kshp=filters_shape,
+                             bsize=batch_size,
+                             border_mode=border_mode,
+                             subsample=subsample,
+                             filters_flip = filters_flip)
+    return conv_op(inputs, filters)
+
+
+
+class BaseAbstractConv2d(Op):
+    """Base class for ConvInferace
+
+    FIXME
+    """
+    check_broadcast = False
+    __props__ = ('border_mode', 'subsample', 'filters_flip', 'imshp', 'kshp', 'bsize')
+
+    def __init__(self,
+                 imshp=None, kshp=None, bsize=None,
+                 border_mode="valid", subsample=(1, 1),
+                 filters_flip = True):
+        if isinstance(border_mode, int):
+            border_mode = (border_mode, border_mode)
+        if isinstance(border_mode, tuple):
+            pad_h, pad_w = map(int, border_mode)
+            border_mode = (pad_h, pad_w)
+        if not ((isinstance(border_mode, tuple) and min(border_mode) >= 0) or
+                border_mode in ('valid', 'full', 'half')):
+            raise ValueError(
+                'invalid border_mode {}, which must be either '
+                '"valid", "full", "half", an integer or a pair of'
+                ' integers'.format(border_mode))
+
+        self.imshp = imshp
+        self.kshp = kshp
+        self.bsize = bsize
+        self.border_mode = border_mode
+        self.filters_flip = filters_flip
+
+        if len(subsample) != 2:
+            raise ValueError("subsample must have two elements")
+        self.subsample = subsample
+
+    def flops(self, inp, outp):
+        """ Useful with the hack in profilemode to print the MFlops"""
+        # if the output shape is correct, then this gives the correct
+        # flops for any direction, sampling, padding, and border mode
+        inputs, filters = inp
+        outputs, = outp
+        assert inputs[1] == filters[1]
```

Comment:
In the gemm-based convolution, something was wrong with this (maybe something was dimshuffled so the assertion won't pass). Haven't read the full diff yet, is this tested anywhere?